vpr/src/route/rr_graph_fpga_interchange.cpp

```diff
@@ -0,0 +1,1706 @@
+     */
+    void create_switch_inf() {
+        std::set<std::tuple<int /*timing*/, bool /*buffered*/>> seen;
+        for (const auto& tile : ar_.getTileTypeList()) {
+            for (const auto& pip : tile.getPips()) {
+                if (pip.isPseudoCells())
+                    continue;
+                seen.emplace(pip.getTiming(), pip.getBuffered21());
+                if (!pip.getDirectional()) {
+                    seen.emplace(pip.getTiming(), pip.getBuffered20());
+                }
+            }
+        }
+        device_ctx_.rr_switch_inf.reserve(seen.size() + 1);
+        int id = 2;
+        make_room_in_vector(&device_ctx_.rr_switch_inf, 0);
+        fill_switch(device_ctx_.rr_switch_inf[RRSwitchId(0)], 0, 0, 0, 0, 0, 0, 0,
+                    vtr::strdup("short"), SwitchType::SHORT);
+        make_room_in_vector(&device_ctx_.rr_switch_inf, 1);
+        fill_switch(device_ctx_.rr_switch_inf[RRSwitchId(1)], 0, 0, 0, 0, 0, 0, 0,
+                    vtr::strdup("generic"), SwitchType::MUX);
+        const auto& pip_models = ar_.getPipTimings();
+        float R, Cin, Cout, Cint, Tdel;
+        std::string switch_name;
+        SwitchType type;
+        for (const auto& value : seen) {
+            make_room_in_vector(&device_ctx_.rr_switch_inf, id);
+            int timing_model_id;
+            int mux_trans_size;
+            bool buffered;
+            std::tie(timing_model_id, buffered) = value;
+            const auto& model = pip_models[timing_model_id];
+            pips_models_[std::make_tuple(timing_model_id, buffered)] = id;
+
+            R = Cin = Cint = Cout = Tdel = 0.0;
+            std::stringstream name;
+            std::string mux_type_string = buffered ? "mux_" : "passGate_";
+            name << mux_type_string;
+
+            R = get_corner_value(model.getOutputResistance(), "slow", "min");
+            name << "R" << std::scientific << R;
+
+            Cin = get_corner_value(model.getInputCapacitance(), "slow", "min");
+            name << "Cin" << std::scientific << Cin;
+
+            Cout = get_corner_value(model.getOutputCapacitance(), "slow", "min");
+            name << "Cout" << std::scientific << Cout;
+
+            if (buffered) {
+                Cint = get_corner_value(model.getInternalCapacitance(), "slow", "min");
+                name << "Cinternal" << std::scientific << Cint;
+            }
+
+            Tdel = get_corner_value(model.getInternalDelay(), "slow", "min");
+            name << "Tdel" << std::scientific << Tdel;
+
+            switch_name = name.str();
+            type = buffered ? SwitchType::MUX : SwitchType::PASS_GATE;
+            mux_trans_size = buffered ? 1 : 0;
+
+            fill_switch(device_ctx_.rr_switch_inf[RRSwitchId(id)], R, Cin, Cout, Cint, Tdel,
+                        mux_trans_size, 0, vtr::strdup(switch_name.c_str()), type);
+
+            id++;
+        }
+    }
```

Comment:
I have seen this function in the `fpga_arch_reader` possibly this can also end up in a utility function of the FPGA interchange?

Reply:
This code exist in read_fpga_interchange.cpp but it works on different data types, extremely similar, but not the same.
This can be unified using templates, but I'm not sure about it

Reply:
It wasn't easy, but in next commit this code will be unified with on from read_fpga_interchage_arch